[0.9.6] influxdb randomly crashing with panic: runtime error: index out of range

```
2015/12/10 00:44:42 InfluxDB starting, version 0.9.6, branch 0.9.6, commit ea8bbabfb68394aab5d9f23f08e29986bc137955, built 2015-12-09T16:10:21+0000
2015/12/10 00:44:42 Go version go1.4.2, GOMAXPROCS set to 12
2015/12/10 00:44:43 Using configuration at: /config/config.toml
[metastore] 2015/12/10 00:44:43 Using data dir: /data/meta
[metastore] 2015/12/10 00:44:43 Skipping cluster join: already member of cluster: nodeId=1 raftEnabled=true peers=[influx-mysql-cp.beget:8088]
[metastore] 2015/12/10 00:44:43 Node at influx-mysql-cp.beget:8088 [Follower]
[metastore] 2015/12/10 00:44:48 Node at influx-mysql-cp.beget:8088 [Leader]. peers=[influx-mysql-cp.beget:8088]
[metastore] 2015/12/10 00:44:48 Updated node id=1 hostname=influx-mysql-cp.beget:8088
[metastore] 2015/12/10 00:44:48 spun up monitoring for 1
[store] 2015/12/10 00:44:48 Using data dir: /data/db
[tsm1wal] 2015/12/10 00:44:48 tsm1 WAL starting with 10485760 segment size
[tsm1wal] 2015/12/10 00:44:48 tsm1 WAL writing to /data/wal/_internal/monitor/2
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00001.wal, size 859
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00004.wal, size 2555
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00005.wal, size 7824
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00007.wal, size 6363
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00009.wal, size 830
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00010.wal, size 831
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00011.wal, size 827
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00012.wal, size 3450
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00013.wal, size 3591
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00015.wal, size 4508
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00016.wal, size 773
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00017.wal, size 6379
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00023.wal, size 4488
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00025.wal, size 8472
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/2/_00040.wal, size 0
[tsm1wal] 2015/12/10 00:44:48 tsm1 WAL starting with 10485760 segment size
[tsm1wal] 2015/12/10 00:44:48 tsm1 WAL writing to /data/wal/_internal/monitor/3
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00001.wal, size 957
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00002.wal, size 3548
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00004.wal, size 5442
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00005.wal, size 781
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00006.wal, size 7095
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00007.wal, size 838
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00009.wal, size 3609
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00014.wal, size 7284
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00015.wal, size 1760
[cacheloader] 2015/12/10 00:44:48 reading file /data/wal/_internal/monitor/3/_00016.wal, size 0
[tsm1wal] 2015/12/10 00:44:48 tsm1 WAL starting with 10485760 segment size
[tsm1wal] 2015/12/10 00:44:48 tsm1 WAL writing to /data/wal/mysql-cp/default/1
[cacheloader] 2015/12/10 00:44:49 reading file /data/wal/mysql-cp/default/1/_00052.wal, size 2381471
[cacheloader] 2015/12/10 00:44:50 reading file /data/wal/mysql-cp/default/1/_00053.wal, size 8730030
[cacheloader] 2015/12/10 00:44:51 reading file /data/wal/mysql-cp/default/1/_00054.wal, size 0
[handoff] 2015/12/10 00:45:24 Starting hinted handoff service
[monitor] 2015/12/10 00:45:24 'hh' registered for diagnostics monitoring
[handoff] 2015/12/10 00:45:24 Using data dir: /data/hh
[subscriber] 2015/12/10 00:45:24 opened service
[monitor] 2015/12/10 00:45:24 Starting monitor system
[monitor] 2015/12/10 00:45:24 'build' registered for diagnostics monitoring
[monitor] 2015/12/10 00:45:24 'runtime' registered for diagnostics monitoring
[monitor] 2015/12/10 00:45:24 'network' registered for diagnostics monitoring
[monitor] 2015/12/10 00:45:24 'system' registered for diagnostics monitoring
[cluster] 2015/12/10 00:45:24 Starting cluster service
[shard-precreation] 2015/12/10 00:45:24 Starting precreation service with check interval of 1m0s, advance period of 1h0m0s
[snapshot] 2015/12/10 00:45:24 Starting snapshot service
[copier] 2015/12/10 00:45:24 Starting copier service
[admin] 2015/12/10 00:45:24 Starting admin service
[monitor] 2015/12/10 00:45:24 Storing statistics in database '_internal' retention policy '', at interval 50s
[admin] 2015/12/10 00:45:24 Listening on HTTP: 192.168.0.36:8083
[continuous_querier] 2015/12/10 00:45:24 Starting continuous query service
[httpd] 2015/12/10 00:45:24 Starting HTTP service
[httpd] 2015/12/10 00:45:24 Authentication enabled: false
[httpd] 2015/12/10 00:45:24 Listening on HTTP: 192.168.0.36:8086
[udp] 2015/12/10 00:45:24 Started listening on UDP: 192.168.0.36:4444
[retention] 2015/12/10 00:45:24 Starting retention policy enforcement service with check interval of 30m0s
[run] 2015/12/10 00:45:24 Listening for signals
2015/12/10 00:45:24 Sending anonymous usage statistics to m.influxdb.com
panic: runtime error: index out of range

goroutine 314 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WriteWALEntry).Encode(0xccd5dda0f8, 0xc9ffe28000, 0x100000, 0x100000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:373 +0x1234
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WAL).writeToLog(0xc208c49100, 0x7f0799e68b40, 0xccd5dda0f8, 0x0, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:208 +0xbb
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WAL).WritePoints(0xc208c49100, 0xc233613080, 0x0, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:149 +0xa3
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*DevEngine).WritePoints(0xc208058960, 0xc31fdec000, 0x3d4, 0x400, 0x0, 0xcba5e16000, 0x3d4, 0x400, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/engine.go:255 +0x55f
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc208c1f340, 0xc31fdec000, 0x3d4, 0x400, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/shard.go:263 +0x66d
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc2080e6000, 0x1, 0xc31fdec000, 0x3d4, 0x400, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/store.go:398 +0x14e
github.com/influxdb/influxdb/cluster.func·002(0x1, 0x1, 0xc31fdec000, 0x3d4, 0x400)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/cluster/points_writer.go:315 +0x131
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/cluster/points_writer.go:347 +0x31a

goroutine 1 [chan receive, 3 minutes]:
main.(*Main).Run(0xc208049700, 0xc20800a010, 0x1, 0x1, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/cmd/influxd/main.go:96 +0x7b3
main.main()
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/cmd/influxd/main.go:46 +0xdc

...
```

influxdb was downloaded from https://s3.amazonaws.com/influxdb/influxdb_0.9.6_x86_64.tar.gz

it seems that influxdb is crashing on writes

after enabling write logs I can see data being written

```
mysql_cp,customer=blond,hostname=tsoi,user=blond_drup2 access_denied=0i,busy_time=0i,commit_transactions=0i,concurrent_connections=0i,connected_time=1i,cpu_time=0i,denied_connections=0i,empty_q
ueries=29i,lost_connections=0i,other_commands=4i,rollback_transactions=0i,rows_fetched=612i,rows_updated=62i,select_commands=128i,table_rows_read=1967i,total_connections=2i,update_commands=101i
 1449709200000000000
mysql_cp,customer=cpaset4v,hostname=tsoi,user=cpaset4v_youne11 access_denied=0i,busy_time=0i,commit_transactions=0i,concurrent_connections=0i,connected_time=0i,cpu_time=0i,denied_connections=0i
,empty_queries=159i,lost_connections=0i,other_commands=19i,rollback_transactions=0i,rows_fetched=2269i,rows_updated=10i,select_commands=274i,table_rows_read=6985i,total_connections=4i,update_co
mmands=11i 1449709200000000000
mysql_cp,customer=tkgery98,hostname=tsoi,user=tkgery98_germes access_denied=0i,busy_time=0i,commit_transactions=0i,concurrent_connections=0i,connected_time=0i,cpu_time=0i,denied_connections=0i,
empty_queries=9i,lost_connections=0i,other_commands=1i,rollback_transactions=0i,rows_fetched=277i,rows_updated=0i,select_commands=15i,table_rows_read=350i,total_connections=1i,update_commands=0
i 1449709200000000000
mysql_cp,customer=efremo8j,hostname=tsoi,user=efremo8j_prodl36 access_denied=0i,busy_time=0i,commit_transactions=0i,concurrent_connections=0i,connected_time=0i,cpu_time=0i,denied_connections=0i
,empty_queries=46i,lost_connections=0i,other_commands=1i,rollback_transactions=0i,rows_fetched=247i,rows_updated=0i,select_commands=261i,table_rows_read=1145i,total_connections=1i,update_comman
ds=0i 1449709200000000000
mysql_cp,customer=mydenzuv,hostname=tsoi,user=mydenzuv_nzen access_denied=0i,busy_time=0i,commit_transactions=0i,concurrent_connections=0i,connected_time=8i,cpu_time=0i,denied_connections=0i,em
pty_queries=27i,lost_connections=0i,other_commands=4i,rollback_transactions=0i,rows_fetched=920i,rows_updated=27i,select_commands=90i,table_rows_read=1211i,total_connections=3i,update_commands=
31i 1449709200000000000
panic: runtime error: index out of range

goroutine 155 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WriteWALEntry).Encode(0xc208904250, 0xc2313ea000, 0x100000, 0x100000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:373 +0x1234
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WAL).writeToLog(0xc2088e7e80, 0x7fe1ac08a5b0, 0xc208904250, 0x0, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:208 +0xbb
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WAL).WritePoints(0xc2088e7e80, 0xc21a3e2ba0, 0x0, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:149 +0xa3
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*DevEngine).WritePoints(0xc208968960, 0xc244208000, 0x351, 0x400, 0x0, 0xc22c32a000, 0x351, 0x400, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/engine.go:255 +0x55f
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc2089386e0, 0xc244208000, 0x351, 0x400, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/shard.go:263 +0x66d
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc2080e4000, 0x1, 0xc244208000, 0x351, 0x400, 0x0, 0x0)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/tsdb/store.go:398 +0x14e
github.com/influxdb/influxdb/cluster.func·002(0x1, 0x1, 0xc244208000, 0x351, 0x400)
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/cluster/points_writer.go:315 +0x131
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
        /tmp/tmp.QrWAFdTR6B/src/github.com/influxdb/influxdb/cluster/points_writer.go:347 +0x31a

```

how to debug this situation?

same crashes began on another 0.9.6 instance (clean install without previous data)

writing dataset is different

```
mysql_quota,customer=isever8k,hostname=rembo,user=isever8k_ivan-mysql blocks=11145216i,files=140i 1449715200000000000
mysql_quota,customer=zin13689,hostname=rembo,user=zin13689_p-mysql blocks=10215424i,files=166i 1449715200000000000
mysql_quota,customer=dimat8,hostname=rembo,user=dimat8_jm2-mysql blocks=23437312i,files=217i 1449715200000000000
mysql_quota,customer=amakse,hostname=rembo,user=amakse_ocart-mysql blocks=13348864i,files=234i 1449715200000000000
mysql_quota,customer=rofanddn,hostname=rembo,user=rofanddn_b-mysql blocks=785924096i,files=1180i 1449715200000000000
mysql_quota,customer=askaro,hostname=rembo,user=askaro_opca-mysql blocks=34291712i,files=410i 1449715200000000000
mysql_quota,customer=grigor0u,hostname=rembo,user=grigor0u_jm1-mysql blocks=3842048i,files=235i 1449715200000000000
mysql_quota,customer=gretany5,hostname=rembo,user=gretany5_prigoto-mysql blocks=2568192i,files=24i 1449715200000000000
panic: runtime error: index out of range

goroutine 382 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WriteWALEntry).Encode(0xc2088ca030, 0xc226b84000, 0x100000, 0x100000, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:373 +0x1234
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WAL).writeToLog(0xc208976000, 0x7f3dc019d280, 0xc2088ca030, 0x0, 0x0, 0x0)
        /go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:208 +0xbb
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*WAL).WritePoints(0xc208976000, 0xc2219f8300, 0x0, 0x0, 0x0)
        /go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:149 +0xa3
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*DevEngine).WritePoints(0xc208978000, 0xc22d202000, 0x1b87, 0x1e00, 0x0, 0xc22c616000, 0x1b87, 0x1c00, 0x0, 0x0)
        /go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/engine.go:255 +0x55f
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc208974000, 0xc22d202000, 0x1b87, 0x1e00, 0x0, 0x0)
        /go/src/github.com/influxdb/influxdb/tsdb/shard.go:263 +0x66d
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc2080fe000, 0x1, 0xc22d202000, 0x1b87, 0x1e00, 0x0, 0x0)
        /go/src/github.com/influxdb/influxdb/tsdb/store.go:398 +0x14e
github.com/influxdb/influxdb/cluster.func·002(0x1, 0x1, 0xc22d202000, 0x1b87, 0x1e00)
        /go/src/github.com/influxdb/influxdb/cluster/points_writer.go:315 +0x131
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
        /go/src/github.com/influxdb/influxdb/cluster/points_writer.go:347 +0x31a

goroutine 1 [chan receive, 1 minutes]:
main.(*Main).Run(0xc208049700, 0xc20800a010, 0x1, 0x1, 0x0, 0x0)
        /go/src/github.com/influxdb/influxdb/cmd/influxd/main.go:96 +0x7b3
main.main()
        /go/src/github.com/influxdb/influxdb/cmd/influxd/main.go:46 +0xdc

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ With this release InfluxDB is moving to Go 1.5.
 - [#4940](https://github.com/influxdb/influxdb/pull/4940): Fix distributed aggregate query query error. Thanks @li-ang
 - [#4622](https://github.com/influxdb/influxdb/issues/4622): Fix panic when passing too large of timestamps to OpenTSDB input.
 - [#5064](https://github.com/influxdb/influxdb/pull/5064): Full support for parenthesis in SELECT clause, fixes [#5054](https://github.com/influxdb/influxdb/issues/5054). Thanks @mengjinglei
+- [#5079](https://github.com/influxdb/influxdb/pull/5079): Ensure tsm WAL encoding buffer can handle large batches.
 
 ## v0.9.6 [2015-12-09]
 

--- a/tsdb/engine/tsm1/cache_test.go
+++ b/tsdb/engine/tsm1/cache_test.go
@@ -354,7 +354,7 @@ func mustMarshalEntry(entry WALEntry) (WalEntryType, []byte) {
 
 	b, err := entry.Encode(bytes)
 	if err != nil {
-		panic("error encoding")
+		panic(fmt.Sprintf("error encoding: %v", err))
 	}
 
 	return entry.Type(), snappy.Encode(b, b)

--- a/tsdb/engine/tsm1/wal.go
+++ b/tsdb/engine/tsm1/wal.go
@@ -366,6 +366,13 @@ func (w *WriteWALEntry) Encode(dst []byte) ([]byte, error) {
 
 	for k, v := range w.Values {
 
+		// Make sure we have enough space in our buf before copying.  If not,
+		// grow the buf.
+		if len(dst[:n])+2+len(k)+len(v)*8+4 > len(dst) {
+			grow := make([]byte, len(dst)*2)
+			dst = append(dst, grow...)
+		}
+
 		switch v[0].Value().(type) {
 		case float64:
 			dst[n] = float64EntryType
@@ -380,19 +387,19 @@ func (w *WriteWALEntry) Encode(dst []byte) ([]byte, error) {
 		}
 		n++
 
-		// Make sure we have enough space in our buf before copying.  If not,
-		// grow the buf.
-		if len(k)+2+len(v)*8+4 > len(dst)-n {
-			grow := make([]byte, len(dst)*2)
-			dst = append(dst, grow...)
-		}
-
 		n += copy(dst[n:], u16tob(uint16(len(k))))
 		n += copy(dst[n:], []byte(k))
 
 		n += copy(dst[n:], u32tob(uint32(len(v))))
 
 		for _, vv := range v {
+
+			// Grow our slice if needed
+			if len(dst[:n])+16 > len(dst) {
+				grow := make([]byte, len(dst)*2)
+				dst = append(dst, grow...)
+			}
+
 			n += copy(dst[n:], u64tob(uint64(vv.Time().UnixNano())))
 			switch t := vv.Value().(type) {
 			case float64:

--- a/tsdb/engine/tsm1/wal.go
+++ b/tsdb/engine/tsm1/wal.go
@@ -365,7 +365,6 @@ func (w *WriteWALEntry) Encode(dst []byte) ([]byte, error) {
 	var n int
 
 	for k, v := range w.Values {
-
 		// Make sure we have enough space in our buf before copying.  If not,
 		// grow the buf.
 		if len(dst[:n])+2+len(k)+len(v)*8+4 > len(dst) {
@@ -393,8 +392,8 @@ func (w *WriteWALEntry) Encode(dst []byte) ([]byte, error) {
 		n += copy(dst[n:], u32tob(uint32(len(v))))
 
 		for _, vv := range v {
-
-			// Grow our slice if needed
+			// Grow our slice if needed. Enough room is needed for the timestamp (8 bytes)
+			// and the value itself (another 8 bytes).
 			if len(dst[:n])+16 > len(dst) {
 				grow := make([]byte, len(dst)*2)
 				dst = append(dst, grow...)

--- a/tsdb/engine/tsm1/wal_test.go
+++ b/tsdb/engine/tsm1/wal_test.go
@@ -70,6 +70,62 @@ func TestWALWriter_WritePoints_Single(t *testing.T) {
 	}
 }
 
+func TestWALWriter_WritePoints_LargeBatch(t *testing.T) {
+	dir := MustTempDir()
+	defer os.RemoveAll(dir)
+	f := MustTempFile(dir)
+	w := tsm1.NewWALSegmentWriter(f)
+
+	var points []tsm1.Value
+	for i := 0; i < 100000; i++ {
+		points = append(points, tsm1.NewValue(time.Unix(int64(i), 0), int64(1)))
+	}
+
+	values := map[string][]tsm1.Value{
+		"cpu,host=A,server=01,foo=bar,tag=really-long#!~#float": points,
+		"mem,host=A,server=01,foo=bar,tag=really-long#!~#float": points,
+	}
+
+	entry := &tsm1.WriteWALEntry{
+		Values: values,
+	}
+
+	if err := w.Write(mustMarshalEntry(entry)); err != nil {
+		fatal(t, "write points", err)
+	}
+
+	if _, err := f.Seek(0, os.SEEK_SET); err != nil {
+		fatal(t, "seek", err)
+	}
+
+	r := tsm1.NewWALSegmentReader(f)
+
+	if !r.Next() {
+		t.Fatalf("expected next, got false")
+	}
+
+	we, err := r.Read()
+	if err != nil {
+		fatal(t, "read entry", err)
+	}
+
+	e, ok := we.(*tsm1.WriteWALEntry)
+	if !ok {
+		t.Fatalf("expected WriteWALEntry: got %#v", e)
+	}
+
+	for k, v := range e.Values {
+		for i, vv := range v {
+			if got, exp := vv.String(), values[k][i].String(); got != exp {
+				t.Fatalf("points mismatch: got %v, exp %v", got, exp)
+			}
+		}
+	}
+
+	if n := r.Count(); n != MustReadFileSize(f) {
+		t.Fatalf("wrong count of bytes read, got %d, exp %d", n, MustReadFileSize(f))
+	}
+}
 func TestWALWriter_WritePoints_Multiple(t *testing.T) {
 	dir := MustTempDir()
 	defer os.RemoveAll(dir)
@@ -490,7 +546,7 @@ func mustMarshalEntry(entry tsm1.WALEntry) (tsm1.WalEntryType, []byte) {
 
 	b, err := entry.Encode(bytes)
 	if err != nil {
-		panic("error encoding")
+		panic(fmt.Sprintf("error encoding: %v", err))
 	}
 
 	return entry.Type(), snappy.Encode(b, b)

